python/cuml/dask/datasets/make_regression.py

```diff
@@ -0,0 +1,216 @@
+                         tail_strength=0.5, random_state=None, n_parts=1):
+    """ Generate a mostly low rank matrix with bell-shaped singular values
+
+    Parameters
+    ----------
+    n_samples : int, optional (default=100)
+        The number of samples.
+    n_features : int, optional (default=100)
+        The number of features.
+    effective_rank : int, optional (default=10)
+        The approximate number of singular vectors required to explain most of
+        the data by linear combinations.
+    tail_strength : float between 0.0 and 1.0, optional (default=0.5)
+        The relative importance of the fat noisy tail of the singular values
+        profile.
+    random_state : int
+        Determines random number generation for dataset creation. Pass an int
+        for reproducible output across multiple function calls.
+    n_parts : int, optional (default=1)
+        The number of parts of work.
+
+    Returns
+    -------
+    X : Dask-cuPy array of shape [n_samples, n_features]
+        The matrix.
+    """
+    rs = da.random.RandomState(seed=random_state,
+                               RandomState=cp.random.RandomState)
+    n = min(n_samples, n_features)
+
+    def generate_chunks_for_qr(total_size, min_size, n_parts):
+
+        n_total_per_part = max(1, int(total_size / n_parts))
+        if n_total_per_part > min_size:
+            min_size = n_total_per_part
+
+        n_partitions = int(max(1, total_size / min_size))
+        rest = total_size % (n_partitions * min_size)
+        chunks_list = [min_size for i in range(n_partitions-1)]
+        chunks_list.append(min_size + rest)
+        return tuple(chunks_list)
+
+    # Random (ortho normal) vectors
+    m1 = rs.standard_normal((n_samples, n),
+                            chunks=(generate_chunks_for_qr(n_samples,
+                                                           n, n_parts), -1))
+    u, _ = da.linalg.qr(m1)
+
+    m2 = rs.standard_normal((n, n_features),
+                            chunks=(-1, generate_chunks_for_qr(n_features,
+                                                               n, n_parts)))
+    v, _ = da.linalg.qr(m2)
+
+    # For final multiplication
+    n_samples_per_part = max(1, int(n_samples / n_parts))
+    u = u.rechunk({0: n_samples_per_part, 1: -1})
+    v = v.rechunk({0: n_samples_per_part, 1: -1})
+
+    # Index of the singular values
+    sing_ind = cp.arange(n, dtype=cp.float64)
+
+    # Build the singular profile by assembling signal and noise components
+    low_rank = ((1 - tail_strength) * cp.exp(-1.0 *
+                                             (sing_ind / effective_rank) **
+                                             2))
+    tail = tail_strength * cp.exp(-0.1 * sing_ind / effective_rank)
+    local_s = low_rank + tail
+    s = da.from_array(local_s,
+                      chunks=(int(n_samples_per_part),), asarray=False)
+
+    return da.dot(u * s, v)
```

Comment:
I'm sure this one is probably okay, but need to verify end-to-end that it's not creating any new arrays on host.

Reply:
I can replace the `da.dot(u * s, v)` by `u *= s` and then `da.dot(u, v)` to do the first multiplication inplace. Unfortunately, the shape of u is (n_samples, n) and the one of v is (n, n_features) where n = min(n_samples, n_features). Because of that there is no guarantee that the resulting matrix of the dot product of shape (n_samples, n_features) can hold into either u or v. Hence, I think we will have to create a copy.

Reply:
Of course, it might be possible to use some trick to hold the result either in u or in v depending on the value of n. But it would look a bit odd.

Reply:
I tried this :

`u *= s`
`if n_samples > n_features:`
`____ u @= v`
`____ return u`
`else:`
 `____ return u @ v`

but it seems like inplace dot product has not been implemented

Reply:
@wxbn, my concern isn't whether this is in place, but rather whether Dask is explicitly using `numpy` under the hood to create any new arrays on host (which can severely hinder performance). It's not a huge deal, but it's something I've come across on a couple occasions.